browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/brave_vpn/dns/brave_vpn_dns_observer_service.h"
+
+#include "chrome/browser/net/secure_dns_config.h"
+#include "chrome/browser/net/secure_dns_util.h"
+#include "chrome/browser/net/stub_resolver_config_reader.h"
+#include "chrome/browser/net/system_network_context_manager.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/browser/ui/browser_window.h"
+#include "chrome/browser/ui/simple_message_box.h"
+#include "chrome/common/pref_names.h"
+#include "chrome/grit/chromium_strings.h"
+#include "components/country_codes/country_codes.h"
+#include "components/grit/brave_components_strings.h"
+#include "components/prefs/pref_service.h"
+#include "net/dns/public/dns_over_https_config.h"
+#include "net/dns/public/doh_provider_entry.h"
+#include "net/dns/public/secure_dns_mode.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace secure_dns = chrome_browser_net::secure_dns;
+
+namespace brave_vpn {
+
+namespace {
+std::string GetFilteredProvidersForCountry() {
+  // Use default servers for current country.
+  auto providers = secure_dns::ProvidersForCountry(
+      secure_dns::SelectEnabledProviders(net::DohProviderEntry::GetList()),
+      country_codes::GetCurrentCountryID());
+  std::string doh_https_templates;
+  for (const auto* entry : providers) {
+    if (!doh_https_templates.empty()) {
+      doh_https_templates += " ";
+    }
+    net::DnsOverHttpsConfig doh_config({entry->doh_server_config});
+    doh_https_templates += doh_config.ToString();
+  }
+  return doh_https_templates;
+}
+std::string GetDoHServers(SecureDnsConfig* dns_config) {
+  return dns_config && !dns_config->doh_servers().servers().empty()
+             ? dns_config->doh_servers().ToString()
+             : GetFilteredProvidersForCountry();
+}
+}  // namespace
+
+BraveVpnDnsObserverService::BraveVpnDnsObserverService(PrefService* local_state)
+    : local_state_(local_state) {
+  pref_change_registrar_.Init(local_state);
+  pref_change_registrar_.Add(
+      prefs::kDnsOverHttpsMode,
+      base::BindRepeating(&BraveVpnDnsObserverService::OnDNSPrefChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+}
+
+BraveVpnDnsObserverService::~BraveVpnDnsObserverService() = default;
+
+bool BraveVpnDnsObserverService::ShouldAllowExternalChanges() const {
+  if (allow_changes_for_testing_.has_value())
```

Comment:
Should this check (and definition for `SetAllowExternalChangesForTesting`) be wrapped with a guard, like `#if !defined(OFFICIAL_BUILD)` (so that we don't accidentally allow changes in a real build)

Reply:
done